Stratis.Bitcoin.Features.BlockStore/LoopSteps/ProcessPendingStorageStep.cs

```diff
@@ -43,132 +42,158 @@ internal ProcessPendingStorageStep(BlockStoreLoop blockStoreLoop, ILoggerFactory
+                return StepResult.Continue;
 
-                context.GetNextBlock();
+            do
+            {
+                PrepareNextBlockFromPendingStorage(context);
 
-                var shouldBreak = ShouldBreakExecution(context);
-                if (shouldBreak)
+                var canProcessNextBlock = context.CanProcessNextBlock();
+                if (canProcessNextBlock == false)
                 {
                     if (context.PendingBlockPairsToStore.Any())
-                        await this.PushBlocksToRepository(context);
+                        await PushBlocksToRepository(context);
 
                     break;
                 }
-            }
+
+                if (context.PendingStorageBatchSize > BlockStoreLoop.MaxPendingInsertBlockSize)
+                {
+                    await PushBlocksToRepository(context);
+
+                    context.PendingBlockPairsToStore.Clear();
+                    context.PendingStorageBatchSize = 0;
+                }
+
+            } while (context.CancellationToken.IsCancellationRequested == false);
 
             return StepResult.Continue;
         }
 
         /// <summary>
-        /// Store missing blocks and remove them from pending blocks and set the Store's tip to <see cref="ProcessPendingStorageContext.NextChainedBlock"/>
+        /// When the node is NOT in IBD, process and push the blocks in <see cref="BlockStoreLoop.PendingStorage"/> immediately 
+        /// to the block repository without checking batch size.
         /// </summary>
-        private async Task PushBlocksToRepository(ProcessPendingStorageContext context)
+        private async Task<StepResult> ProcessWhenNotInIBD(ProcessPendingStorageContext context)
         {
-            await this.BlockStoreLoop.BlockRepository.PutAsync(context.PendingBlockPairsToStore.First().ChainedBlock.HashBlock, context.PendingBlockPairsToStore.Select(b => b.Block).ToList());
-            this.BlockStoreLoop.SetStoreTip(context.PendingBlockPairsToStore.First().ChainedBlock);
+            this.logger.LogDebug("{0}:'{1}/{2}'", nameof(context.NextChainedBlock), context.NextChainedBlock.HashBlock, context.NextChainedBlock.Height);
+
+            do
+            {
+                PrepareNextBlockFromPendingStorage(context);
+
+                await PushBlocksToRepository(context);
+
+                var canProcessNextBlock = context.CanProcessNextBlock();
+                if (canProcessNextBlock == false)
+                    break;
 
-            this.logger.LogTrace("({0}:{1} / {2}.{3}:{4} / {5}:{6}')", nameof(context.BlockStoreLoop.ChainState.IsInitialBlockDownload), context.BlockStoreLoop.ChainState.IsInitialBlockDownload, nameof(context.PendingBlockPairsToStore), nameof(context.PendingBlockPairsToStore.Count), context.PendingBlockPairsToStore?.Count, nameof(context.PendingStorageBatchSize), context.PendingStorageBatchSize);
+            } while (context.CancellationToken.IsCancellationRequested == false);
```

Comment:
No logical reason to use do;while as we are only checking IsCancellationRequested so better check it before rather then after.

Reply:
OK, I had a look through the code you wrote first, way back when, and there we did first check the cancellation token before processing or pushing anything to the repo, I've swopped this around.